stats/cloud/collector.go

```diff
@@ -196,41 +202,54 @@ func (c *Collector) Link() string {
 // at regular intervals and when the context is terminated.
 func (c *Collector) Run(ctx context.Context) {
 	wg := sync.WaitGroup{}
-
+	quit := make(chan struct{})
 	// If enabled, start periodically aggregating the collected HTTP trails
 	if c.config.AggregationPeriod.Duration > 0 {
 		wg.Add(1)
 		aggregationTicker := time.NewTicker(time.Duration(c.config.AggregationCalcInterval.Duration))
 
 		go func() {
+			defer wg.Done()
 			for {
 				select {
+				case <-c.stopSendingMetricsCh:
+					return
 				case <-aggregationTicker.C:
 					c.aggregateHTTPTrails(time.Duration(c.config.AggregationWaitPeriod.Duration))
 				case <-ctx.Done():
 					c.aggregateHTTPTrails(0)
 					c.flushHTTPTrails()
-					c.pushMetrics()
-					wg.Done()
+					close(quit)
 					return
 				}
 			}
 		}()
 	}
 
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for range c.pushMetricsCh {
+			c.pushMetrics()
+		}
+	}()
+
 	defer func() {
+		close(c.pushMetricsCh)
 		wg.Wait()
 		c.testFinished()
 	}()
 
 	pushTicker := time.NewTicker(time.Duration(c.config.MetricPushInterval.Duration))
 	for {
 		select {
-		case <-pushTicker.C:
-			c.pushMetrics()
-		case <-ctx.Done():
-			c.pushMetrics()
+		case <-c.stopSendingMetricsCh:
+			return
```

Comment:
Again here is the possibility for the select to select one of the other cases if they also can proceed ... and given that the specification specifically says that it's at random that's a problem ;). So I still think that this should be in a select of it's own .. and that below we should just call pushMetrics instead of having another separate goroutine ...
Thinking about it ... you definitely should not be doing it in this way with another goroutine because what we want is for there to be only one pushMetrics at a time but if the previous one takes more than the MetricPushInterval there will be two pushes at the same time ... possibly leading to closing the c.stopSendingMetricsCh .. twice... 


Also not that I don't think you should remove it but `c.pushMetricsCh` could've been a local variable instead of a field on the struct

Reply:
Not sure what you mean. Structure this way, we definitely only allow 1 `c.pushMetrics()` run at a time, so the code is safe.